Scrollbar appearance for page and for textarea do not update correctly

lilyspiniolas opened the pull request:
#### 5661743448a09e46a1794643d1868694ce1f8f27
<pre>
Scrollbar appearance for page and for textarea do not update correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=293160">https://bugs.webkit.org/show_bug.cgi?id=293160</a>
<a href="https://rdar.apple.com/151496190">rdar://151496190</a>

Reviewed by Aditya Keerthi.

`useDarkAppearance` is used during compositing geometry updates for each
`ScrollableArea` to update the scrollbar appearance accordingly. However,
compositing geometry updates were not previously guaranteed to happen when
the value of the color-scheme style property changed. We now call
`setNeedsCompositingGeometryUpdate` after updating appearance so that the
scrollbars immediately display with the new appearance.

A layout test has not been added as a repaint occurs when the test finishes
by default, which causes a reference test to always pass, even without these
changes. When the final repaint was disabled, a false fail occurred. When
a repaint test was used instead to compare repaint rects, a false pass occurred.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::recalculateScrollbarOverlayStyle):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::updateScrollbarsAfterStyleChange):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:

Canonical link: <a href="https://commits.webkit.org/295590@main">https://commits.webkit.org/295590@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1a1899c091acbac7d3276c567114249071f26803

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105350 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25062 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15489 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110558 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56002 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107391 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25500 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33604 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80014 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108356 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95081 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60323 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13161 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55402 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13203 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113236 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32506 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23987 "") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89093 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32866 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91301 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88736 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33622 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11413 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27975 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17119 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32428 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37899 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32197 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35543 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33773 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/rendering/RenderLayerScrollableArea.h:
```diff
@@ -323,6 +328,8 @@ class RenderLayerScrollableArea final : public ScrollableArea, public CanMakeChe
     ScrollPosition m_scrollPosition;
     std::optional<ScrollPosition> m_postLayoutScrollPosition;
 
+    Style::Color m_backgroundColor { Color::white };
```
I don't think we should have a `Style::Color` member outside of `/style`. Could this be a `Color`?

pxlcoder commented on Source/WebCore/rendering/RenderLayerScrollableArea.h:
```diff
@@ -315,6 +318,8 @@ class RenderLayerScrollableArea final : public ScrollableArea, public CanMakeChe
     
     bool m_isRegisteredForAnimatedScroll { false };
 
+    bool m_hasDarkBackground { false };
```
I don't think this member is necessary, since it can be derived from `m_backgroundColor`?

pxlcoder commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -1606,7 +1630,12 @@ bool RenderLayerScrollableArea::isHandlingWheelEvent() const
 
 bool RenderLayerScrollableArea::useDarkAppearance() const
```
Changing the return value of `useDarkAppearance` itself is a little concerning, because this has always followed color-scheme, and could affect other things.

Why does this need to be changed, and not just `useDarkAppearanceForScrollbars()`?

lilyspiniolas replied:
It's not required for the intended behavior change to work, updated to remove the override.

pxlcoder commented on LayoutTests/scrollbars/scrollbar-appearance-updates-immediately.html:
```diff
@@ -0,0 +1,25 @@
+        setTimeout(function() {
+            document.body.style.colorScheme = "dark";
+            testRunner.notifyDone();
+        }, 1000);
```
Do we need to use `1000` or would `0` work?

lilyspiniolas replied:
tested just now, `0` works. will update

lilyspiniolas committed:
Scrollbar appearance for page and for textarea do not update correctly
https://bugs.webkit.org/show_bug.cgi?id=293160
rdar://151496190

Reviewed by Aditya Keerthi.

`useDarkAppearance` is used during compositing geometry updates for each
`ScrollableArea` to update the scrollbar appearance accordingly. However,
compositing geometry updates were not previously guaranteed to happen when
the value of the color-scheme style property changed. We now call
`setNeedsCompositingGeometryUpdate` after updating appearance so that the
scrollbars immediately display with the new appearance.

A layout test has not been added as a repaint occurs when the test finishes
by default, which causes a reference test to always pass, even without these
changes. When the final repaint was disabled, a false fail occurred. When
a repaint test was used instead to compare repaint rects, a false pass occurred.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::recalculateScrollbarOverlayStyle):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::updateScrollbarsAfterStyleChange):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:

Canonical link: https://commits.webkit.org/295590@main
